include/sparrow/dynamic_bitset.hpp

```diff
@@ -380,4 +562,249 @@ namespace sparrow
+    }
+    
+    template <class B>
+    auto bitset_reference<B>::operator|=(bool rhs) noexcept -> self_type&
+    {
+        if (rhs)
+        {
+            set();
+        }
+        return *this;
+    }
+    
+    template <class B>
+    auto bitset_reference<B>::operator^=(bool rhs) noexcept -> self_type&
+    {
+        if (rhs)
+        {
+            bool old_value = m_block & m_mask;
+            m_block ^= m_mask;
+            m_bitset.update_null_count(old_value, !old_value);
+        }
+        return *this;
+    }
+
+    template <class B>
+    bitset_reference<B>::bitset_reference(bitset_type& bitset, block_type& block, block_type mask)
+        : m_bitset(bitset)
+        , m_block(block)
+        , m_mask(mask)
+    {
+    }
+
+    template <class B>
+    void bitset_reference<B>::assign(bool rhs) noexcept
+    {
+        rhs ? set() : reset();
+    }
+
+    template <class B>
+    void bitset_reference<B>::set() noexcept
+    {
+        bool old_value = m_block & m_mask;
+        m_block |= m_mask;
+        m_bitset.update_null_count(old_value, m_block & m_mask);
+    }
+
+    template <class B>
+    void bitset_reference<B>::reset() noexcept
+    {
+        bool old_value = m_block & m_mask;
+        m_block &= ~m_mask;
+        m_bitset.update_null_count(old_value, m_block & m_mask);
+    }
+
+    template <class B1, class B2>
+    bool operator==(const bitset_reference<B1>& lhs, const bitset_reference<B2>& rhs)
+    {
+        return bool(lhs) == bool(rhs);
+    }
+
+    /**********************************
+     * bitset_iterator implementation *
+     **********************************/
+
+    template <class B, bool is_const>
+    bitset_iterator<B, is_const>::bitset_iterator(bitset_type* bitset, block_type* block, size_type index)
+        : p_bitset(bitset)
+        , p_block(block)
+        , m_index(index)
+    {
+        assert(m_index < bitset_type::s_bits_per_block);
+    }
+    
+    template <class B, bool is_const>
+    auto bitset_iterator<B, is_const>::dereference() const -> reference
+    {
+        if constexpr (is_const)
+        {
+            return (*p_block) & bitset_type::bit_mask(m_index);
+        }
+        else
+        {
+            return bitset_reference<B>(*p_bitset, *p_block, bitset_type::bit_mask(m_index));
+        }
+    }
+
+    template <class B, bool is_const>
+    void bitset_iterator<B, is_const>::increment()
+    {
+        ++m_index;
+        // If we have reached next block
+        if (is_first_bit_of_block(m_index))
+        {
+            ++p_block;
+            m_index = 0u;
+        }
+        assert(m_index < bitset_type::s_bits_per_block);
+    }
+
+    template <class B, bool is_const>
+    void bitset_iterator<B, is_const>::decrement()
+    {
+        // decreasing moves to previous block
+        if (is_first_bit_of_block(m_index))
+        {
+            --p_block;
+            m_index = sizeof(block_type);
+        }
+        else
+        {
+            --m_index;
```

Comment:
Is this what we want when `m_index` is zero?
Note that the valeu would then switch to the highest index, and that's *defined behavior*. But is it valid for us? Should we assert here?
Or should `m_index` be changed to a difference type instead (which would then be signed)?

Reply:
If `m_index` is 0, then `is_first_bit_of_block(m_index)` is true and we never reach the `else`.